bindings/sundials4py/idas/idas.cpp

```diff
@@ -0,0 +1,319 @@
+      fn_table->MEMBER = nb::cast(fn);                                             \
+      if (fn) { return NAME(ida_mem, which, &WRAPPER); }                           \
+      else { return NAME(ida_mem, which, nullptr); }                               \
+    },                                                                             \
+    __VA_ARGS__)
+
+#define BIND_IDAB_CALLBACK2(NAME, FN_TYPE1, MEMBER1, WRAPPER1, FN_TYPE2, \
+                            MEMBER2, WRAPPER2, ...)                      \
+  m.def(                                                                 \
+    #NAME,                                                               \
+    [](void* ida_mem, int which,                                         \
+       std::function<std::remove_pointer_t<FN_TYPE1>> fn1,               \
+       std::function<std::remove_pointer_t<FN_TYPE2>> fn2)               \
+    {                                                                    \
+      void* user_data   = nullptr;                                       \
+      auto fn_table     = get_idasa_fn_table(ida_mem, which);            \
+      fn_table->MEMBER1 = nb::cast(fn1);                                 \
+      fn_table->MEMBER2 = nb::cast(fn2);                                 \
+      if (fn1) { return NAME(ida_mem, which, WRAPPER1, WRAPPER2); }      \
+      else { return NAME(ida_mem, which, nullptr, WRAPPER2); }           \
+    },                                                                   \
+    __VA_ARGS__)
+
+namespace sundials4py {
+
+void bind_idas(nb::module_& m)
+{
+#include "idas_generated.hpp"
+
+  nb::class_<IDAView>(m, "IDAView")
+    .def("get", nb::overload_cast<>(&IDAView::get, nb::const_),
+         nb::rv_policy::reference);
+
+  m.def(
+    "IDASetOptions",
+    [](void* ida_mem, const std::string& idaid, const std::string& file_name,
+       int argc, const std::vector<std::string>& args)
+    {
+      std::vector<char*> argv;
+
+      for (const auto& arg : args)
+      {
+        // We need a non-const char*, so we use data() and an explicit cast.
+        // This is safe as long as the underlying std::string is not modified.
+        argv.push_back(const_cast<char*>(arg.data()));
+      }
+
+      return IDASetOptions(ida_mem, idaid.empty() ? nullptr : idaid.c_str(),
+                           file_name.empty() ? nullptr : file_name.c_str(),
+                           argc, argv.data());
+    },
+    nb::arg("ida_mem"), nb::arg("idaid"), nb::arg("file_name"), nb::arg("argc"),
+    nb::arg("args"));
+
+  m.def(
+    "IDACreate",
+    [](SUNContext sunctx)
+    { return std::make_shared<IDAView>(IDACreate(sunctx)); },
+    nb::arg("sunctx"), nb::keep_alive<0, 1>());
+
+  m.def("IDAInit",
+        [](void* ida_mem, std::function<std::remove_pointer_t<IDAResFn>> res,
+           sunrealtype t0, N_Vector yy0, N_Vector yp0)
+        {
+          int ida_status = IDAInit(ida_mem, idas_res_wrapper, t0, yy0, yp0);
```

Comment:
Since we always pass `ida_res_wrapper` to `IDAInit` check if the residual is `NULL` and check return flag
```suggestion
          if (!res) { throw sundials4py::illegal_value("res was null"); }
          int ida_status = IDAInit(ida_mem, idas_res_wrapper, t0, yy0, yp0);
          if (ida_status != IDA_SUCCESS)
          {
            throw sundials4py::error_returned("Failed to initialize IDA");
          }
```
Should we return the error flag?